### PR TITLE
First draft of integrated documentation for Documents endpoint with CRUD

### PR DESCRIPTION
**Pull request by monotasker**
This version of the Documents endpoint docs includes POST, PUT, and DELETE methods. I've also done some editing in the existing GET documentation.

Two things to note. First, the old documentation was inconsistent about whether the endpoint was "document" or "documents." I regularized this to the singular, but I can easily change it all to plural if that's desired.

Second, in some places the automerge doesn't seem to have lined up equivalent sections of the document. So you may still want to go through and look at the proposed version manually.

### CODE REVIEW

File: Documents-Endpoint.md [hunk 1]
```diff
@@ -17,48 +28,56 @@ The Documents endpoint is used to access the data for documents, as opposed to m
...
-| start | (For range) Start of a range of passages (can't be used with `ref`) | GET |
-| end |  (For range) End of a range of passages (requires `start` and no `ref`) | GET |
+### Validating Changes to Documents
 
-### Response Headers
+Because of its flexibility, at some points this specification relies on clients to ensure that the XML they submit is correctly structured. This requires more than simply that clients submit valid TEI XML when creating or editing a text segment. It is strongly recommended that servers implementing these extended methods offer robust validation of any submitted XML. This should include validating that the submitted text or fragment
 
-The response contains the following response headers:
+- is well-formed XML
+- satsfies the requirements of the TEI schema
```
Comment (mromanello) [inline Documents-Endpoint.md]: typo: satfies => satisfies

File: Documents-Endpoint.md [hunk 2]
```diff
@@ -249,19 +364,491 @@ Retrieve the full document bgu;11;2029
...
+
+#### Successful response headers
+
+The response headers after a successful POST Documents request should include a `Location` value. This should be the URL where the newly inserted text segment(s) can be retrieved via a GET request. The response should also include a `Content-type` header with a value of "application/tei+xml".
+
+
+The response should also include a `Link` header as detailed in the core documentation for the Documents endpoint. This `Link` gives URL references for the previous and next segments of the document, a URL for the document's navigation structure, and a URL for the document's Collections metadata. __If no previous or next segment exists, those parts of the `Link` header should give the URL for the newly created segment instead.__ If the request has created multiple segments, the `next` value for the `Link` header should be the last created segment, and the `previous` value should be the first created segment.
```
Comment (balmas) [inline Documents-Endpoint.md]: "__If no previous or next segment exists, those parts of the `Link` header should give the URL for the newly created segment instead.__ " I don't understand this. Why would we return prev/next links in this case?

File: Documents-Endpoint.md [hunk 3]
```diff
@@ -249,19 +364,491 @@ Retrieve the full document bgu;11;2029
...
+
+#### Successful response body
+
+The response body after a successful POST request should contain an XML object representing the newly created text segment (and any children). This should be identical to the response a client would receive using a GET request to the `Location` indicated in the response header. This will also mean that in most successful requests the XML contained in the response body will be identical to the object sent by the client in the POST.
+
+#### Unsuccessful response headers
+
+In an unsuccessful request, the response headers should include a `Location` whose value is the URL for the Documents endpoint documentation. The response should also include a `Content-type` header with a value of "application/tei+xml".
```
Comment (balmas) [inline Documents-Endpoint.md]: this seems a little unusual. Is it a standard approach?

File: Documents-Endpoint.md [hunk 4]
```diff
@@ -249,19 +364,491 @@ Retrieve the full document bgu;11;2029
...
+
+### PUT Request Body
+
+The body of the `PUT` request must be wrapped in a TEI rootnode containing a `<dts:fragment>` entity. See the further specifications under "Default Request and Response Body Format" above.
+
+The contents of this `<dts:fragment>` must be a single XML element (along with its enclosed text and/or children) representing the modified form of the target document section. This single element inside the `<dts:fragment>` will replace the corresponding XML element in the document on the server.   
+
+The text/XML submitted in the body of a Document's `PUT` request will completely replace the XML entity representing the identified text segment. So the submitted fragment __must include the outer element__ identified by the "ref" value. If you identify your submitted text as a modified version of line "12", and if each line is represented by a TEI `<ln>` element, you would include the opening and closing tags `<ln xml:id="16">` and `</ln>` around the changed content. This is to allow modification of the attributes on the outer element tag.
```
Comment (balmas) [inline Documents-Endpoint.md]: should the example `<ln xml:id="16">` be `<ln xml:id="12">` ?

File: Documents-Endpoint.md [hunk 5]
```diff
@@ -249,19 +364,491 @@ Retrieve the full document bgu;11;2029
...
+          <rdg wit="#Bertalotto #p">እማኅደሩ</rdg>
+        </app>
+    </div>
+  </dts:fragment>
+</TEI>
+```
+
+##### Successful PUT response status
+
+- `201(Created)`
```
Comment (balmas) [inline Documents-Endpoint.md]: should be 200(OK) response code

File: Documents-Endpoint.md [hunk 6]
```diff
@@ -1,13 +1,24 @@
-# Documents Endpoint
+# Document Endpoint
 
-The Documents endpoint is used to access the data for documents, as opposed to metadata (which is found in collections).  The representation of a document is up to the implementation.
+The documents endpoint is used to access and modify the content of documents, as opposed to metadata (which is found in collections). Implementations must at least support reading documents using GET requests as described below. The other request methods (POST, PUT, DELETE) are not required and they may be implemented selectively. For example, an implementation could allow PUT operations to update the contents of a document, but not support the total removal of document sections using DELETE. if an implementation is going to allow modification of document contents please do so using these methods as described below. The API documentation for each implementation must clearly indicate which methods are supported for this endpoint.
```
[the remark this answers is not in the document]
Comment (PonteIneptique) [reply Documents-Endpoint.md]: Isn't that the point of OPTIONS in http queries ?

File: Documents-Endpoint.md (hunk 4 above)
Comment (hcayless) [reply Documents-Endpoint.md]: And there's no `<ln>` element in TEI. `<l>`?

File: Documents-Endpoint.md (hunk 1 above)
Comment (monotasker) [reply Documents-Endpoint.md]: Fixed. Thanks.

File: Documents-Endpoint.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (monotasker) [reply Documents-Endpoint.md]: Fixed. Thanks.

[the remark this answers is not in the document]
Comment (monotasker) [reply Documents-Endpoint.md]: I've added a basic reference to supportedOperations. I think we should flesh this out some, but I need to figure some things out about hydra first.

File: Documents-Endpoint.md [hunk 7]
```diff
@@ -17,48 +28,56 @@ The Documents endpoint is used to access the data for documents, as opposed to m
...
-| last | Last passage of the document in the Documents endpoint |
-| contents | Link to the Navigation Endpoint for the current document |
-| collections | Link to the Collections endpoint for the current document |
+This specification leaves the choice to enforce editing restrictions like this up to the implementer. If such restrictions are imposed, any request that is rejected as a result must be accompanied by a clear error message. The error message must (a) explain the restriction, and (b) provide a URL where the restriction is documented.
+
+## URI
+
+### Query Parameters
+
+The documents endpoint supports the following query parameters:
```
[the remark this answers is not in the document]
Comment (monotasker) [reply Documents-Endpoint.md]: Fixed. Thanks.

File: Documents-Endpoint.md [hunk 8]
```diff
@@ -90,29 +109,118 @@ Here is a template of the URI for Documents API. The route itself (`/dts/api/doc
...
   ]
 }
 ```
 
-## Examples
+## GET Requests on the Document Endpoint
+
+### GET query parameters
+
+The only strictly required parameter for a GET documents request is `id`. If neither `before` nor `after` is supplied, the server should interpret the request as supplying the initial form of a new document. In that case, the request should fail if (a) some text already exists for that document, or (b) the request body contains a `<dts:fragment>` element.
```
[the remark this answers is not in the document]
Comment (monotasker) [reply Documents-Endpoint.md]: Fixed. Thanks.

File: Documents-Endpoint.md [hunk 9]
```diff
@@ -249,19 +364,491 @@ Retrieve the full document bgu;11;2029
...
+
+The POST method of the Documents endpoint allows for creation of new textual passages in a resource. __This method requires that the document already has a metadata record accessible via the Collections endpoint__.
+
+Note that this method should __not__ be used to modify existing segments of text. In other words, if the specified reference label already exists in the document on the server, this method should return an error and refuse to perform the requested operation. If, for example, a document already has a line 6 following the existing line 5, then POST cannot be used to insert new or added text in that existing line 6. Modification of the text in existing document segments must be done using the PUT method instead.
+
+### POST Query parameters
+
+The only strictly required parameters for a POST Documents request are `id` and (if supported) `token`. If neither `before` nor `after` is supplied, the server should interpret the request as supplying the initial form of a new document. In that case, the request should fail if (a) some text already exists for that document, or (b) the request body contains a `<dts:fragment>` element.
```
[the remark this answers is not in the document]
Comment (monotasker) [reply Documents-Endpoint.md]: Fixed. Thanks.

File: Documents-Endpoint.md [hunk 10]
```diff
@@ -249,19 +364,491 @@ Retrieve the full document bgu;11;2029
...
+
+In most cases, though, the request will be inserting a new structural segment into a document that already contains some segments. In such instances, the query must include either a `before` or an `after` parameter to indicate where the new text segment will be inserted.
+
+**Neither the `ref` parameter nor the pair of `start`/`end` parameters should be provided in a POST Documents request. The reference information for the inserted segment(s) should be included using the standard TEI attributes in the XML of the request body.**
```
[the remark this answers is not in the document]
Comment (monotasker) [reply Documents-Endpoint.md]: I've tried to clarify this now.

File: Documents-Endpoint.md (hunk 2 above)
Comment (monotasker) [reply Documents-Endpoint.md]: We agreed to remove this use of the Link header.

File: Documents-Endpoint.md (hunk 3 above)
Comment (monotasker) [reply Documents-Endpoint.md]: This was my misunderstanding. I think that hydra recommends using a Link header to reference documentation instead. So for the moment I've changed it to that.

File: Documents-Endpoint.md (hunk 4 above)
Comment (monotasker) [reply Documents-Endpoint.md]: Fixed. Thanks.

**Pull request merged**